modules/navier_stokes/src/actions/NSFVAction.C

```diff
@@ -0,0 +1,2289 @@
+
+    _problem->addFVKernel(kernel_type, "pins_energy_diffusion", params);
+  }
+  else
+  {
+    const std::string kernel_type = "FVDiffusion";
+    InputParameters params = _factory.getValidParams(kernel_type);
+    params.set<NonlinearVariableName>("variable") = NS::T_fluid;
+    params.set<std::vector<SubdomainName>>("block") = _blocks;
+    params.set<MooseFunctorName>("coeff") = _thermal_conductivity_name;
+
+    _problem->addFVKernel(kernel_type, "ins_energy_diffusion", params);
+  }
+}
+
+void
+NSFVAction::addINSEnergyAmbientConvection()
+{
+  unsigned int num_convection_blocks = _ambient_convection_blocks.size();
+  unsigned int num_used_blocks = num_convection_blocks ? num_convection_blocks : 1;
+
+  const std::string kernel_type = "PINSFVEnergyAmbientConvection";
+  InputParameters params = _factory.getValidParams(kernel_type);
+  params.set<NonlinearVariableName>("variable") = NS::T_fluid;
+  params.set<MooseFunctorName>(NS::T_fluid) = NS::T_fluid;
+  params.set<bool>("is_solid") = false;
+
+  for (unsigned int block_i = 0; block_i < num_used_blocks; ++block_i)
+  {
+    std::string block_name = "";
+    if (num_convection_blocks)
+    {
+      params.set<std::vector<SubdomainName>>("block") = {_ambient_convection_blocks[block_i]};
+      block_name = _ambient_convection_blocks[block_i];
+    }
+    else
+      params.set<std::vector<SubdomainName>>("block") = _blocks;
+
+    params.set<MooseFunctorName>("h_solid_fluid") = _ambient_convection_alpha[block_i];
+    params.set<MooseFunctorName>(NS::T_solid) = _ambient_temperature[block_i];
+
+    _problem->addFVKernel(kernel_type, "ambient_convection_" + block_name, params);
+  }
+}
+
+void
+NSFVAction::addINSEnergyExternalHeatSource()
+{
+  const std::string kernel_type = "FVCoupledForce";
+  InputParameters params = _factory.getValidParams(kernel_type);
+  params.set<NonlinearVariableName>("variable") = NS::T_fluid;
+  params.set<std::vector<SubdomainName>>("block") = _blocks;
+  params.set<CoupledName>("v") = getParam<CoupledName>("external_heat_source");
+  params.set<Real>("coef") = getParam<Real>("external_heat_source_coeff");
+
+  _problem->addFVKernel(kernel_type, "external_heat_source", params);
+}
+
+void
+NSFVAction::addScalarAdvectionKernels()
+{
+  for (const auto & vname : _passive_scalar_names)
+  {
+    const std::string kernel_type = "INSFVScalarFieldAdvection";
+    InputParameters params = _factory.getValidParams(kernel_type);
+    params.set<NonlinearVariableName>("variable") = vname;
+    params.set<MooseEnum>("velocity_interp_method") = "rc";
+    params.set<MooseEnum>("advected_interp_method") = _passive_scalar_advection_interpolation;
+
+    if (_porous_medium_treatment)
+      params.set<UserObjectName>("rhie_chow_user_object") = "pins_rhie_chow_interpolator";
+    else
+      params.set<UserObjectName>("rhie_chow_user_object") = "ins_rhie_chow_interpolator";
+
+    _problem->addFVKernel(kernel_type, "ins_" + vname + "_advection", params);
```

Comment:
I think you could write a lot more kernels like this for porous_medium_treatment. Would reduce the size of the action for kernels by like 40%

Reply:
I personally find it easier to read if the code doesn't have an if statement every other line. There are a lot of things that can change based on the porous medium treatment (rc-object name, velocity names, kernel name, few input parameters). But you are right, some of the kernel functions could be much simpler without the lack of readability, I'll collapse these.